File: backend/internal/models/topic.go
package models

type Topic struct {
	ID            int    `json:"id"`
	Name          string `json:"name"`
	Description   string `json:"description"`
	NoOfPosts     int    `json:"no_of_posts"`
	NoOfFollowers int    `json:"no_of_followers"`
	IsFollowing   bool   `json:"is_following"`
}

type UserTopic struct {
	UserID  int `json:"user_id"`
	TopicID int `json:"topic_id"`
}

type FollowTopicRequest struct {
	IsFollow bool `json:"is_follow" schema:"is_follow"`
}

type ListTopicsRequest struct {
	Page            int    `json:"page,omitempty" schema:"page"`
	PageSize        int    `json:"page_size,omitempty" schema:"page_size"`
	Sort            string `json:"sort,omitempty" schema:"sort"`
	OrderBy         string `json:"order_by,omitempty" schema:"order_by"`
	Search          string `json:"search,omitempty" schema:"search"`
	FilterFollowing bool   `json:"filter_following,omitempty" schema:"filter_following"`
}

File: frontend/src/types/topic.ts
import type z from "zod";
import type {
  listTopicsRequestSchema,
  listTopicsSearchParamsSchema,
  topicsSortValueSchema,
} from "@/schema/topics";

export interface Topic {
  id: number;
  name: string;
  description: string;
  no_of_posts: number;
  no_of_followers: number;
  is_following: boolean;
}

export type TopicsSortValue = z.infer<typeof topicsSortValueSchema>;

export type ListTopicsSearchParams = z.infer<typeof listTopicsSearchParamsSchema>;

export type ListTopicsRequest = z.infer<typeof listTopicsRequestSchema>;

export type FollowTopicRequest = {
  topic_slug: string;
  is_follow: boolean;
};

export interface PaginatedTopicsResponse {
  topics: Topic[];
  count: number;
}
